python_modules/libraries/dagster-openai/dagster_openai/resources.py

```diff
@@ -0,0 +1,234 @@
+        add_to_asset_metadata(context, usage_metadata, output_name)
+
+        return response
+
+    return wrapper
+
+
+class OpenAIResource(ConfigurableResource):
+    """This resource is wrapper over the
+    `openai library <https://github.com/openai/openai-python>`_.
+
+    By configuring this OpenAI resource, you can interact with OpenAI API
+    and log its usage metadata in the asset metadata.
+
+    Examples:
+        .. code-block:: python
+
+            import os
+
+            from dagster import AssetExecutionContext, Definitions, EnvVar, asset, define_asset_job
+            from dagster_openai import OpenAIResource
+
+
+            @asset(compute_kind="OpenAI")
+            def openai_asset(context: AssetExecutionContext, openai: OpenAIResource):
+                with openai.get_client(context) as client:
+                    client.chat.completions.create(
+                        model="gpt-3.5-turbo",
+                        messages=[{"role": "user", "content": "Say this is a test"}]
+                    )
+
+            openai_asset_job = define_asset_job(name="openai_asset_job", selection="openai_asset")
+
+            defs = Definitions(
+                assets=[openai_asset],
+                jobs=[openai_asset_job],
+                resources={
+                    "openai": OpenAIResource(api_key=EnvVar("OPENAI_API_KEY")),
+                },
+            )
+    """
+
+    api_key: str = Field(description=("OpenAI API key. See https://platform.openai.com/api-keys"))
+
+    _client: Client = PrivateAttr()
+
+    @classmethod
+    def _is_dagster_maintained(cls) -> bool:
+        return True
+
+    def _wrap_with_usage_metadata(
+        self,
+        api_endpoint_class: ApiEndpointClassesEnum,
+        context: AssetExecutionContext,
+        output_name: Optional[str],
+    ):
+        for attribute_names in API_ENDPOINT_CLASSES_TO_ENDPOINT_METHODS_MAPPING[
+            api_endpoint_class
+        ]:
+            curr = self._client.__getattribute__(api_endpoint_class.value)
+            # Get the second to last attribute from the attribute list to reach the method.
+            i = 0
+            while i < len(attribute_names) - 1:
+                curr = curr.__getattribute__(attribute_names[i])
+                i += 1
+            # Wrap the method.
+            curr.__setattr__(
+                attribute_names[i],
+                with_usage_metadata(
+                    context=context,
+                    output_name=output_name,
+                    func=curr.__getattribute__(attribute_names[i]),
+                ),
+            )
+
+    def setup_for_execution(self, context: InitResourceContext) -> None:
+        # Set up an OpenAI client based on the API key.
+        self._client = Client(api_key=self.api_key)
+
+    @contextmanager
+    def get_client(
+        self,
+        context: Union[AssetExecutionContext, OpExecutionContext],
+        output_name: Optional[str] = None,
```

Comment:
note on unit tests, since you have output_name here: worth testing the behavior in the following cases:
1. multi_assets (https://docs.dagster.io/concepts/assets/multi-assets)
2. assets with partitions (https://docs.dagster.io/concepts/partitions-schedules-sensors/partitioning-assets)
3. graph-backed assets (https://docs.dagster.io/concepts/assets/graph-backed-assets) - i doubt this will behave any different than a regular asset but worth throwing a test case for complexity

Reply:
All done in [dagster-openai/dagster_openai_tests/test_resources.py](https://github.com/dagster-io/dagster/pull/19697/files#diff-db7e37302fefbb53ccf043549f955c4aea4c4ecc275257ee6dc163c88059a5d1)